File: OJ/leetcode/python/52_n-queens_ii.py
def totalNQueens(n):
    def isValid(board, col, diagonal_down, diagonal_up, i, j):
        return col[j] and diagonal_up[i + j] and \
            diagonal_down[len(col) - i - 1 + j]

    def dfs(filling, i, board, col, diagonal_down, diagonal_up):
        if i == len(col):
            if filling == 0:
                return 1
            else:
                return 0
        
        count = 0
        for j in range(len(col)):
            if isValid(board, col, diagonal_down, diagonal_up, i, j):
                board[i][j] = 'Q'
                col[j] = 0
                diagonal_up[i + j] = 0
                diagonal_down[len(col) - i - 1 + j] = 0
                count += dfs(filling - 1, i + 1, board, col, diagonal_down, diagonal_up)
                diagonal_down[len(col) - i - 1 + j] = 1
                diagonal_up[i + j] = 1
                col[j] = 1
        return count

    board = [['.' for _ in range(n)] for _ in range(n)]
    col = [1 for _ in range(n)]
    diagonal_down = [1 for _ in range(2 * n - 1)]
    diagonal_up = [1 for _ in range(2 * n - 1)]
    return dfs(n, 0, board, col, diagonal_down, diagonal_up)


print(totalNQueens(1))
print(totalNQueens(2))
print(totalNQueens(3))
print(totalNQueens(4))
print(totalNQueens(5))
print(totalNQueens(6))
print(totalNQueens(7))
print(totalNQueens(8))
print(totalNQueens(9))

File: OJ/leetcode/cpp/52_n-queens_ii.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool isValid(int x, int y, 
             const vector<bool>& col,
             const vector<bool>& down,
             const vector<bool>& up) {
    return col[y] and up[x + y] and down[col.size() - x - 1 + y];
}

int dfs(int x, vector<bool>& col, vector<bool>& down, vector<bool>& up,
        const int n) {
    if (x == n) {
        return 1;
    }
    int count = 0;
    for (int y = 0; y < col.size(); ++y) {
        if (isValid(x, y, col, down, up)) {
            col[y] = false;
            down[n - x - 1 + y] = false;
            up[x + y] = false;
            count += dfs(x + 1, col, down, up, n);
            up[x + y] = true;
            down[n - x - 1 + y] = true;
            col[y] = true;
        }
    }
    return count;
}

int totalNQueens(int n) {
    vector<bool> col(n, true);
    vector<bool> down(2 * n - 1, true);
    vector<bool> up(2 * n - 1, true);
    return dfs(0, col, down, up, n);
}

int main() {
    for (int n = 1; n < 10; ++n) {
        cout << totalNQueens(n) << "\n";
    }
    return 0;
}